### DISCUSSION TITLE
about cystal speed of 328p and 328d

### DISCUSSION THREAD

**Discussion by youxiaojie** [dbuezas/lgt8fx#132] [General]
dear friend, I see the official mini evb's schematic, I found that they take 12Mhz for 328p,andd 11.0592Mhz for larduino (328d?), such strange number. does it have effect to our program? or if I take a common 16Mhz to repalce them? if I compile ino, how I select the frequency from menu? and wemos xi has place to add a crystal, what speed is best? 32Mhz 32.768Mhz?16Mhz?

**Comment by LaZsolt** [comment]
With this new boards.txt (which modified by @jg1uaa and @LaZsolt), you could use that board which assembled with a 12 MHz crystal:
https://github.com/dbuezas/lgt8fx/blob/7f6a0c2bf4404459fd3f981d4c32f7afcd4314fa/lgt8f/boards.txt
For use of 11.0592 MHz you must edit the boards.txt. (Duplicate 12 MHz section and change 12 to 11 and recalculate frequencies where is need.) When using 11.0592 MHz little timing errors should be expected. But this package on that board will work.
(New bootloader not necessary.)

**Reply by youxiaojie** [reply]
I notice a question about 32.768K crystal!
1.dear, friend, I see the the datasheet, 328p and d both support 32.768k crystal, and I have a board leaving a pad, and seemly ready for a 32.768K crystal by size.
and how to set this board?

ver 1.05 p20
LGT8FX8D 内部 OSC 振荡器可以工作在高频和低频两种模式下,用户需要根据外接晶振
的实际大小控制内部 OSC 振荡器工作在正确的模式下。同样内部的 RC 振荡器也分为高频和
低频两种。PMCR 寄存器的最低 4 位用于控制这四种时钟源。控制关系如下:

lgt328d internal osc can run in high frequency mode and low frency mod, to setting the register to insure internal osc work in suitable mod. the same to internal RC. the 4bit LSB does this:
PMCR
PMCR[0] 32M RC 1 enable 0 disable
PMCR[1] 32K RC 1 enable 0 disable
PMCR[2]400K ~ 20MHz OSC RC 1 enable 0 disable
PMCR[3]32K ~ 400K OSC 1 enable 0 disable

LGT8FX8D 系统上电后,默认使用 32MHz RC 作为系统时钟源,内核工作在时钟源的 8 分
频(2MHz)。用户可以通过设置 PMCR 寄存器以及系统预分频寄存器(CLKPR)改变默认配置。
如果用户需要更改主时钟源配置,需要在切换时钟前保证切换后的时钟源处于稳定的工
作状态。因此需要在切换主时钟源之前,通过 PMCR[3:0]使能所需时钟源,并等待到时钟稳
定后才能进行切换。
lgt328 using 32M RC as default, core works at 32Mhz/8=4Mhz(? error on datasheet!),we can setting PMCR and CLKPR tomodify default. setting PMCR[3:0] to enable clock source, then after stable, PMCR[6:5] switch main clock source.

时钟源使能并等待稳定后,可以通过 PMCR[6:5]切换主时钟。其中 PMCR[5]用于选择是
内部 RC 振荡器和外部晶振,PMCR[6]用于选择高速时钟源和低速时钟源。

主时钟源选择:
PMCR[6] PMCR[5] 主时钟源
0 0 内部 32MHz RC 振荡器(系统默认)internal hfrc default
0 1 外部 400K ~ 20MHz 高速晶振 external hfosc
1 0 内部 32KHz RC 振荡器 in'l rc
1 1 外部 32K ~ 400KHz 低速晶振 ex'l osc

So I want to ask here,does our lib can handle with pmcr[3:0]=1011 pmcr[6;5]=11?

and another question
from 328p datasheet: (v105 p22)
E2P_clk
E2P_clk 时钟用于产生 FLASH 接口访问时序。E2P_clk 产生访问 E2PCTL 访问 FLASH 接口
的时序。(E2P_clk 固定来自内部 32MHz HFRC 振荡器的 32 分频(1MHz)。如果用户需要使
用 E2PCTL 模块读写内部程序 FLASH 或者数据 FLASH 空间,需要提前使能内部 32MHz 振
荡器。e2p_clk is fixed from in'l 32Mhz hfrc divided by 32,(32Mhz/32), we need to enable 32Mhz hfRC before using e2pctl model r/w in'l program space and data space,)

however 328d is not mention this.

therefor I thought the eeprom lib recently tested. If I using external osc, is it work correct?
and therefor I ask this question:
So I want to ask here,does our lib can handle with pmcr[3:0]=1011 pmcr[6;5]=11?
it contains two meaning, one is about 32.768Khz crystal, the other is for compatible and safe, we enable and change to externel osc for accurate clock while keep int'l 32M RC and 32k RC enabled at the same time?

and bout CLKPR, is it suitable for ex'l crystal? no document how to set up when using ex'l crystal of 32.768khz. do you have interest to test it?

**Reply by youxiaojie** [reply]
> With this new boards.txt (which modified by @jg1uaa and @LaZsolt), you could use that board which assembled with a 12 MHz crystal:
> https://github.com/dbuezas/lgt8fx/blob/7f6a0c2bf4404459fd3f981d4c32f7afcd4314fa/lgt8f/boards.txt
> For use of 11.0592 MHz you must edit the boards.txt. (Duplicate 12 MHz section and change 12 to 11 and recalculate frequencies where is need.) When using 11.0592 MHz little timing errors should be expected. But this package on that board will work.
> (New bootloader not necessary.)
http://ruemohr.org/~ircjunk/avr/baudcalc/avrbaudcalc-1.0.8.php
there is a baud calculater is it suitable for this chip?

**Comment by ithinkido** [comment]
The boards I recently received are using a 16Mhz external crystal.
![IMG_20220215_140149](https://user-images.githubusercontent.com/52834821/154067529-8e36572f-23de-460d-82cc-adce866d9960.jpg)

 I took a look at the new boards.txt, but was not able to find any reference to the crystal speed as previously in the old board.txt defined [here](https://github.com/dbuezas/lgt8fx/blob/7f6a0c2bf4404459fd3f981d4c32f7afcd4314fa/lgt8f/boards.txt)

This is the section of the current boards.txt in the current master branch.
```
# Clock source
328.menu.clock_source.internal=Internal
328.menu.clock_source.internal.build.clock_source=1
328.menu.clock_source.external=External (assumes 32MHz crystal)
328.menu.clock_source.external.build.clock_source=2

# Clock frequencies
328.menu.clock.32MHz=32 MHz
328.menu.clock.32MHz.build.f_cpu=32000000L
328.menu.clock.16MHz=16 MHz
328.menu.clock.16MHz.build.f_cpu=16000000L
328.menu.clock.8MHz=8 MHz
328.menu.clock.8MHz.build.f_cpu=8000000L
328.menu.clock.4MHz=4 MHz
328.menu.clock.4MHz.build.f_cpu=4000000L
328.menu.clock.2MHz=2 MHz
328.menu.clock.2MHz.build.f_cpu=2000000L
328.menu.clock.1MHz=1 MHz
328.menu.clock.1MHz.build.f_cpu=1000000L
```
Does anyone know the correct way to add in a definition here for a 16Mhz external crystal that will still allow a 32Mhz (is this even possible?) processor clock speed to the new boards.txt definitions?

Looking back at this thread, it would be a good idea to add a few more options for external crystal values.

**Reply by LaZsolt** [reply]
Maybe this works with 16 or 12 MHz. https://github.com/dbuezas/lgt8fx/blob/fb4f06a097e6a225a20d1b78d7e70eac160a6d5b/lgt8f/boards.txt

If you want other frequencies see and learn in this file the `LGT8F328 + 12 MHz crystal` section first.
After this you may create your own section with your crystal speed. I hope so.
